diff -Naur 5.0/usr/share/jolla-settings/pages/wlan/mainpage.qml 5.0_mine/usr/share/jolla-settings/pages/wlan/mainpage.qml
--- 5.0/usr/share/jolla-settings/pages/wlan/mainpage.qml	2025-09-12 07:49:25.656822346 +0200
+++ 5.0_mine/usr/share/jolla-settings/pages/wlan/mainpage.qml	2025-09-12 07:56:23.405938668 +0200
@@ -16,6 +16,40 @@
     property bool showAddNetworkDialog
     property bool pageReady
 
+    property string codeString
+
+    function parseCodeString(code) {
+        /* See also: https://github.com/zxing/zxing/wiki/Barcode-Contents
+         *
+         * 'wifi:S:sApname;T:sType;P:sPassword;H:bHidden;I:sEAPIdentity;H:bHidden;E:sEAPMethod;PH2:sEAPPhase2'
+         * S: SSID
+         * T: WEP or WPA or WPA2-EAP or nopass
+         * E: TTLS or PWD
+         *
+         */
+        // NB: "code" is an url, not a string
+        var scheme = /^wifi:/i;
+        var escaped = /\\[\\;,":]/g; // may be backslash-escaped
+        var tokens = code.toString().replace(scheme,'').split(';');
+        var codes = {}
+        for (var i=0; i < tokens.length; ++i) {
+            var pair = tokens[i].split(':')
+            if (pair.length < 2) continue; // empty value --> undefined
+            if (pair && pair[1] && pair[1].length > 0) codes[pair[0]] = pair[1].replace(escaped, '');
+        }
+        var nwconfig = {
+            ssid:         codes["S"],
+            securityType: codes["T"],
+            passphrase:   codes["P"],
+            anon:         codes["A"],
+            identity:     codes["I"],
+            hidden:       (codes["H"] === "true"),
+            eapMethod:    codes["E"],
+            phase2:       codes["PH2"]
+        }
+        return nwconfig
+    }
+
     onStatusChanged: {
         if (status == PageStatus.Active) {
             pageReady = true
@@ -25,16 +59,45 @@
                 var addNetworkProperties = networkHelper.readSettings()
                 var dialog = pageStack.push(Qt.resolvedUrl("AddNetworkDialog.qml"),
                                             {}, PageStackAction.Immediate)
-                dialog.network.ssid = addNetworkProperties.ssid
-                dialog.network.hidden = addNetworkProperties.hidden
-                dialog.network.securityType = addNetworkProperties.securityType
-                if (addNetworkProperties.eapMethod !== undefined)
-                    dialog.network.eapMethod = addNetworkProperties.eapMethod
-                if (addNetworkProperties.phase2)
-                    dialog.network.phase2 = addNetworkProperties.phase2
-                dialog.network.identity = addNetworkProperties.identity
-                dialog.network.passphrase = addNetworkProperties.passphrase
-                if (addNetworkProperties.caCert) {
+                var providedSettings
+                if (codeString !== undefined)
+                    providedSettings = parseCodeString(codeString)
+                if (providedSettings !== undefined) {
+                    dialog.network.ssid = providedSettings.ssid
+                    dialog.network.hidden = providedSettings.hidden
+                    dialog.network.passphrase = providedSettings.passphrase
+                    switch ( providedSettings.securityType ) {
+                        case "nopass":   dialog.network.securityType = NetworkService.SecurityNone; break;
+                        case "WEP":      dialog.network.securityType = NetworkService.SecurityWEP; break;
+                        case "WPA":      dialog.network.securityType = NetworkService.SecurityPSK; break;
+                        case "WPA2-EAP": dialog.network.securityType = NetworkService.SecurityIEEE802
+                                         dialog.network.identity     = providedSettings.identity
+                                         // not supported by dialog?
+                                         dialog.network.anonymousIdentity = providedSettings.anon
+                                         switch (providedSettings.eapMethod) {
+                                             case "TLS":  dialog.network.eapMethod = NetworkService.EapTLS; break;
+                                             case "TTLS": dialog.network.eapMethod = NetworkService.EapTTLS; break;
+                                             case "PEAP": dialog.network.eapMethod = NetworkService.EapPEAP; break;
+                                             case "PWD":  dialog.network.eapMethod = NetworkService.EapPEAP; break;
+                                             default:     dialog.network.eapMethod = NetworkService.EapNone; break;
+                                         }
+                                         dialog.network.anonymousIdentity = providedSettings.anon
+                                         break;
+                        default:         dialog.network.securityType = NetworkService.SecurityPSK; break;
+                    }
+                    // FIXME: add EAP parameters as well
+                } else {
+                    dialog.network.ssid = addNetworkProperties.ssid
+                    dialog.network.hidden = addNetworkProperties.hidden
+                    dialog.network.securityType = addNetworkProperties.securityType
+                    dialog.network.passphrase = addNetworkProperties.passphrase
+                    if (addNetworkProperties.eapMethod !== undefined)
+                        dialog.network.eapMethod = addNetworkProperties.eapMethod
+                    if (addNetworkProperties.phase2)
+                        dialog.network.phase2 = addNetworkProperties.phase2
+                    dialog.network.identity = addNetworkProperties.identity
+                }
+                 if (addNetworkProperties.caCert) {
                     dialog.network.caCert = addNetworkProperties.caCert
                 } else if (addNetworkProperties.caCertFile) {
                     dialog.network.caCertFile = addNetworkProperties.caCertFile
diff -Naur 5.0/usr/share/jolla-settings/settings.qml 5.0_mine/usr/share/jolla-settings/settings.qml
--- 5.0/usr/share/jolla-settings/settings.qml	2025-09-12 07:51:18.066377181 +0200
+++ 5.0_mine/usr/share/jolla-settings/settings.qml	2025-09-12 07:51:44.914882461 +0200
@@ -99,6 +99,16 @@
                                     { "showAddNetworkDialog": true })
         }
 
+        function showAddNetworkDialogFromQR(code) {
+            const scheme = /^wifi:/i;
+            if (!scheme.test(code.toString())) { // "code" is an url, not a string
+                showAddNetworkDialog()
+            } else {
+                console.info("Adding Wi-Fi network from QR code")
+                window.showSettingsPage("system_settings", Qt.resolvedUrl("pages/wlan/mainpage.qml"), { "showAddNetworkDialog": true, "codeString": code })
+            }
+        }
+
         function importVpn(args) {
             window.showSettingsPage("system_settings", Qt.resolvedUrl("pages/vpn/mainpage.qml"),
                                     { "importPath": args[1], "importMime": args[0] })
diff -Naur 5.0/usr/share/jolla-settings/settings.qml.orig 5.0_mine/usr/share/jolla-settings/settings.qml.orig
--- 5.0/usr/share/jolla-settings/settings.qml.orig	1970-01-01 01:00:00.000000000 +0100
+++ 5.0_mine/usr/share/jolla-settings/settings.qml.orig	2025-09-12 07:51:26.550532123 +0200
@@ -0,0 +1,184 @@
+import QtQuick 2.0
+import Sailfish.Silica 1.0
+// Load translations
+import com.jolla.settings 1.0
+import Nemo.DBus 2.0
+import Nemo.Notifications 1.0
+import "./pages"
+
+ApplicationWindow {
+    id: window
+
+    property Page _mainPage
+
+    allowedOrientations: defaultAllowedOrientations
+    _defaultPageOrientations: Orientation.All
+    _defaultLabelFormat: Text.PlainText
+
+    initialPage: Component { MainPage { id: mainPage; Component.onCompleted: window._mainPage = mainPage} }
+
+    cover: Qt.resolvedUrl("pages/SettingsCover.qml")
+
+    // Function to show a setting page specified by the url. This cleans the stack
+    // to the first page before pushing the new pages on a stack. Otherwise the stack
+    // might have the same page twice.
+    function showSettingsPage(section, url, properties) {
+        if (!pageStack.currentPage.backNavigation) {
+            showBusyWarning()
+            return
+        }
+
+        pageStack.pop(_mainPage, PageStackAction.Immediate)
+        _mainPage.moveToSection(section)
+        pageStack.push(url, properties || {}, PageStackAction.Immediate)
+        window.activate()
+    }
+
+    function showSettingsSection(section) {
+        if (!pageStack.currentPage.backNavigation) {
+            showBusyWarning()
+            return
+        }
+
+        pageStack.pop(_mainPage, PageStackAction.Immediate)
+        _mainPage.moveToSection(section)
+        window.activate()
+    }
+
+    function showBusyWarning() {
+        //: system notification shown when settings is requested to show a page but it's busy
+        //% "Settings application is busy"
+        notification.previewBody = qsTrId("settings-he-warning_settings_busy")
+        notification.publish()
+    }
+
+    Notification {
+        id: notification
+
+        isTransient: true
+    }
+
+    // Add new signals and signal handlers below if there's a need to support opening
+    // other pages too.
+    DBusAdaptor {
+        service: "com.jolla.settings"
+        path: "/com/jolla/settings/ui"
+        iface: "com.jolla.settings.ui"
+
+        function showSettings() {
+            window.showSettingsSection("system_settings")
+        }
+
+        function showTransfers() {
+            window.showSettingsPage("system_settings", Qt.resolvedUrl("pages/transferui/mainpage.qml"))
+        }
+
+        function showAccounts() {
+            window.showSettingsSection("accounts")
+        }
+
+        function showSailfishOs() {
+            window.showSettingsPage("system_settings", Qt.resolvedUrl("pages/sailfishos/mainpage.qml"))
+        }
+
+        function showLocationSettings() {
+            window.showSettingsPage("system_settings", Qt.resolvedUrl("pages/gps_and_location/location.qml"))
+        }
+
+        function showEventsSettings() {
+            window.showSettingsPage("system_settings", Qt.resolvedUrl("pages/events/events.qml"))
+        }
+
+        function showCallRecordings() {
+            window.showSettingsPage("applications", Qt.resolvedUrl("pages/jolla-voicecall/voicecall.qml"),
+                                    { "showRecordingsImmediately": true })
+        }
+
+        function showAddNetworkDialog() {
+            window.showSettingsPage("system_settings", Qt.resolvedUrl("pages/wlan/mainpage.qml"),
+                                    { "showAddNetworkDialog": true })
+        }
+
+        function importVpn(args) {
+            window.showSettingsPage("system_settings", Qt.resolvedUrl("pages/vpn/mainpage.qml"),
+                                    { "importPath": args[1], "importMime": args[0] })
+        }
+
+        function findBluetoothDevices() {
+            window.showSettingsPage("system_settings", Qt.resolvedUrl("pages/bluetooth/bluetoothSettings.qml"))
+            pageStack.currentPage.autoStartDiscovery()
+        }
+
+        function newVpnConnection() {
+            window.showSettingsPage("system_settings", Qt.resolvedUrl("pages/vpn/mainpage.qml"))
+            pageStack.push(Qt.resolvedUrl("pages/vpn/NewConnectionDialog.qml"), {}, PageStackAction.Immediate)
+        }
+
+        function showAmbienceSettings(ambienceContentId) {
+            window.showSettingsPage("system_settings", Qt.resolvedUrl("pages/jolla-gallery-ambience/ambience.qml"), {})
+            pageStack.push("com.jolla.gallery.ambience.AmbienceSettingsPage",
+                           { "contentId": ambienceContentId }, PageStackAction.Immediate)
+        }
+
+        function importWebcal(path) {
+            var comp = Qt.createComponent("/usr/share/accounts/ui/webcal.qml")
+            if (comp.status !== Component.Ready) {
+                console.warn("Cannot load webcal.qml: " + comp.errorString())
+                return
+            }
+            var agent = comp.createObject(window, {'remoteUrl': path})
+            if (agent === null) {
+                console.warn("Unable to instantiate webcal.qml")
+                return
+            }
+            showAccounts()
+            if (pageStack.currentPage.backNavigation) {
+                // We're not busy and can show the account page.
+                agent.endDestination = pageStack.currentPage
+                agent.endDestinationAction = PageStackAction.Pop
+                pageStack.push(agent.initialPage, {}, PageStackAction.Immediate)
+            }
+        }
+
+        function addNewUser() {
+            window.showSettingsPage("system_settings", Qt.resolvedUrl("pages/users/users.qml"),
+                                    { "enableUserCreationOnceComplete": true })
+        }
+
+        function showPage(path) {
+            path = path.toString()
+            var obj = _mainPage.objectForPath(path)
+            if (obj) {
+                if (obj.type === "page") {
+                    var params = obj.data()["params"]
+
+                    if (params["source"]) {
+                        window.showSettingsPage(path, params["source"])
+                    } else {
+                        console.warn("Settings app requested to show settings page, but no page source defined for the config '"
+                                     + path + "'")
+                    }
+                } else {
+                    console.warn("Settings app requested to show a settings page, but the config '"
+                                 + path + "' is of wrong type '" + obj.type + "'")
+                }
+            } else {
+                console.warn("Settings app requested to show settings page, but no config found for path", path)
+            }
+        }
+
+        function activateWindow(arg) {
+            window.activate()
+        }
+    }
+
+    DBusAdaptor {
+        path: "/share_signing_keys"
+        iface: "org.sailfishos.share"
+
+        function share(shareActionConfiguration) {
+            window.showSettingsPage("system_settings", Qt.resolvedUrl("pages/keys/SigningSharePage.qml"),
+                                    { "shareActionConfiguration": shareActionConfiguration })
+        }
+    }
+}
